src/main/scala/Lst.scala

```diff
@@ -0,0 +1,282 @@
+package dogs
+
+import cats._
+import cats.implicits._
+import scala.annotation.tailrec
+
+/**
+ * Immutable, singly-linked list implementation.
+ *
+ * This code is very similar to scala.List, with a few key differences:
+ *
+ * 1. It does not expose any "unsafe" methods.
+ * 2. It is invariant, whereas scala.List is covariant.
+ * 3. It uses subtyping to differentiate non-emptiness.
+ * 4. It provides Cats type class instances.
+ * 5. It does not currently provide a fast (mutable) builder.
+ *
+ * The types defined here are as follows:
+ *
+ *  - Lst[A] represents a list of zero-or-more elements.
+ *  - Nel[A] represents a list of one-or-more elements.
+ *  - El[A]  represents an empty list (exactly zero elements).
+ *
+ * (Every Lst[A] is either a Nel[A] or an El[A].)
+ *
+ * While it does not (yet) provide every single Scala collection
+ * method, it provides a decent number of them. Type class instances
+ * are provided for both Lst and Nel (for example, Nel has a Comonad
+ * instance, whereas Lst only has a CoflatMap instance).
+ */
+sealed abstract class Lst[A] {
+
+  final def isEmpty: Boolean =
+    this match {
+      case El() => true
+      case _ => false
+    }
+
+  final def cata[B](b: => B, f: (A, Lst[A]) => B): B =
+    this match {
+      case Nel(h, t) => f(h, t)
+      case El() => b
+    }
+
+  final def getNonEmpty: Option[Nel[A]] =
+    this match {
+      case nel: Nel[_] => Some(nel)
+      case El() => None
+    }
+
+  final def ::(a: A): Nel[A] =
+    Nel(a, this)
+
+  @tailrec final def foldLeft[B](b: B)(f: (B, A) => B): B =
+    this match {
+      case El() => b
+      case Nel(h, t) => t.foldLeft(f(b, h))(f)
+    }
+
+  def foldRight[B](lb: Eval[B])(f: (A, Eval[B]) => Eval[B]): Eval[B]
+
+  def :::(as: Lst[A]): Lst[A]
```

Comment:
In general, I prefer using pattern matching, but I used delegation anywhere that the return type for the `Nel` and `Lst` versions are different.

If you'd prefer consistency, I could make all of the methods use delegation.